Articles/FileSystemProviders/extending.md

```diff
@@ -0,0 +1,213 @@
+---
+v8-equivalent: "https://our.umbraco.com/documentation/Extending/FileSystemProviders/"
+versionFrom: 9.0.0
+verified-against: beta-1
+---
+
+# Custom file systems (IFileSystem)
+
+## Media Filesystem
+By default, Umbraco uses an instance of `PhysicalFileSystem` to handle the storage location of the media archive (wwwroot/media).
+
+This can be configured during composition:
+
+```csharp
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Umbraco.Cms.Core.Composing;
+using Umbraco.Cms.Core.DependencyInjection;
+using Umbraco.Cms.Core.Hosting;
+using Umbraco.Cms.Core.IO;
+using Umbraco.Cms.Infrastructure.DependencyInjection;
+
+namespace UmbracoExamples.Composition
+{
+    public class SetMediaFileSystemComposer : IUserComposer
+    {
+        public void Compose(IUmbracoBuilder builder)
+        {
+            builder.SetMediaFileSystem((factory) =>
+            {
+                IHostingEnvironment hostingEnvironment = factory.GetRequiredService<IHostingEnvironment>();
+                var folderLocation = "~/CustomMediaFolder";
+                var rootPath = hostingEnvironment.MapPathWebRoot(folderLocation);
+                var rootUrl = hostingEnvironment.ToAbsolute(folderLocation);
+
+                return new PhysicalFileSystem(
+                    factory.GetRequiredService<IIOHelper>(),
+                    hostingEnvironment,
+                    factory.GetRequiredService<ILogger<PhysicalFileSystem>>(),
+                    rootPath,
+                    rootUrl);
+            });
+        }
+    }
+```
+
+When creating a PhyscialFileSystem it takes some dependencies like IIOHelper, but the last two parameters are what we're interested in. The rootPath is where your media will be stored on the disk, since netcore by default store files in the wwwroot, we must put our desired folder somewhere within wwwroot, to ensure that we use `hostingEnvironment.MapPathWebRoot(~/CustomMediaFolder)`, the ~ will be mapped to your wwwroot folder, so the final rootPath will be your/project/path/wwwroot/CustomMediaFolder, the ~ is therefore important!
+
+The other part of the puzzle is the rootUrl, which is the base URL your media files will be served from, so in this case, your image URL could look something like mysite.com/CustomMediaFolder/MyAwesomePicture.png, again the ~ is important. Another thing worth mentioning with the rootUrl is that it must be the same as the folder location, otherwise, you will get 404's for your images.
```

Comment:
```suggestion
The other part of the puzzle is the `rootUrl`, which is the base URL your media files will be served from. In this case, your image URL could look something like `mysite.com/CustomMediaFolder/MyAwesomePicture.png`. Again the `~` is important. Another thing worth mentioning with the `rootUrl` is that it must be the same as the folder location, otherwise, you will get 404's for your images.
```